tests/federation_room_event_auth_test.go

```diff
@@ -0,0 +1,206 @@
+package tests
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/matrix-org/gomatrixserverlib"
+	"github.com/tidwall/gjson"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/federation"
+	"github.com/matrix-org/complement/internal/must"
+)
+
+func TestInboundFederationRejectsEventsWithRejectedAuthEvents(t *testing.T) {
+	/* These tests check that events which refer to rejected events in auth_events
+	 * are themselves rejected.
+	 *
+	 * In order to inject an outlier, we include it as an extra auth_event in a
+	 * regular event. Doing so means that the regular event should itself be
+	 * rejected.
+	 *
+	 * We finish up by sending a final, normal, event which should be accepted
+	 * everywhere. This acts as a sentinel so that we can be sure that the
+	 * events have all been correctly propagated.
+	 *
+	 * The DAG ends up looking like this:
+	 *
+	 *       C
+	 *     / | \
+	 *    /  R  \
+	 *   |   ^   \
+	 *   |    .. O
+	 *   |       ^
+	 *   X .......
+	 *   |
+	 *   S
+	 *
+	 * Where:
+	 *    .... represents an "auth_event" link
+	 *    C is the room creation series
+	 *    R is a rejected event
+	 *    O is an outlier, which should be rejected
+	 *    X is an event with O among its auth_events, which should be rejected
+	 *           as a side-effect of O being rejected
+	 *    S is the final regular event, which acts as a sentinel
+	 *
+	 * To check if the outlier is rejected, we simply request the event via
+	 * /rooms/{roomID}/event. If it is rejected, we should get a 404.
```

Comment:
Sanity check: spec says that 404 means `The event was not found or you do not have permission to read this event.` Is it possible that 404 could be returned even if the event was accepted? Don't think so---we'have permission to see (at least) all events after we joined.

Reply:
yes, I don't think permissions should be a factor here. We're using a `public_chat` preset which makes history visible to everyone anyway.